c++/src/kj/exception.c++

```diff
@@ -830,24 +830,69 @@ void Exception::addTrace(void* ptr) {
   }
 }
 
+#if !KJ_NO_EXCEPTIONS
+
+namespace {
+
+KJ_THREADLOCAL_PTR(ExceptionImpl) currentException = nullptr;
+
+}  // namespace
+
 class ExceptionImpl: public Exception, public std::exception {
 public:
-  inline ExceptionImpl(Exception&& other): Exception(mv(other)) {}
+  inline ExceptionImpl(Exception&& other): Exception(mv(other)) {
+    insertIntoCurrentExceptions();
+  }
   ExceptionImpl(const ExceptionImpl& other): Exception(other) {
     // No need to copy whatBuffer since it's just to hold the return value of what().
+    insertIntoCurrentExceptions();
+  }
+  ~ExceptionImpl() {
+    // Look for ourselves in the list.
+    for (auto* ptr = &currentException; *ptr != nullptr; ptr = &(*ptr)->nextCurrentException) {
+      if (*ptr == this) {
+        *ptr = nextCurrentException;
+        return;
+      }
+    }
+
+    // Possibly the ExceptionImpl was destroyed on a different thread than created it? That's
+    // pretty bad, we'd better abort.
```

Comment:
This seems like a surprising constraint to me. I could imagine needing be able to move exceptions across threads... but I suppose the way to do that is to return one from an executeAsync() callback, which will make sure the objects are constructed and destroyed on their respective threads?

Reply:
Note that `Exception`s can still be moved between threads just fine. Only `ExceptionImpl` cannot. Instances of `ExceptionImpl` are only constructed for the purpose of throwing an exception, and I'm reasonably certain the C++ exception-throwing machinery won't destroy a thrown object on a different thread than the one that threw it.